### DISCUSSION TITLE
It takes 2 seconds to automatically switch to visual mode when opening the sublime

### DISCUSSION THREAD

**Discussion by xianyuIC** [NeoVintageous/NeoVintageous#959] [Q&A]
My sublime adopts the default mode of NeoVintageous, which defaults to visual mode. But when I open a file in Terminal, it starts in insert mode and takes 2 seconds for it to automatically change to visual mode. If operate quickly. For example, when I open a new file with sublime and press G to jump to the end of the file, it becomes typing the G character.

But if I have already opened a Sublime and then opened another file with sublime in Terminal, it will not have a 2-second inset mode, but will directly be in visual mode

Please help to see if this is a bug and can it be fixed?

**Comment by gerardroche** [comment]
There is no default visual mode. Or you're using the `default-visual-mode` branch? That is not included in the latest release. I just rebased the branch on top of the latest release.

I can't reproduce issue.

If you could get some logs: https://github.com/NeoVintageous/NeoVintageous/blob/master/.github/CONTRIBUTING.md#debugging

There might be some errors in the console that will tell us more.

**Comment by gerardroche** [comment]
It's ok. I think I can reproduce it. When the file is first open it doesn't seem to enter visual mode at all. The character the cursor is on looks like a visual selection, but in nv it seems to think it's in insert mode.

**Comment by gerardroche** [comment]
Ok. Try that. I think I was correct. It was in visual mode but missing a flag to indicate to nv that it was actually visual mode.

**Reply by gerardroche** [reply]
ahh, i think there's an off-by-one issue now.

**Comment by gerardroche** [comment]
It looks like there is an off-by-one issues on commands like goto definition e.g. `CTRL-]`.

**Reply by xianyuIC** [reply]
Can this issue be fixed?

**Reply by gerardroche** [reply]
Should be fixed now. 

The goto definition off-by-one issue only happens if the file has not been previously loaded, but there are probably other cases where it is an issue too.

I had to solve it by change the on_activated() event to on_activated_async(), but this may not be a viable solution going forward. It fixes, what looks like, an underlying ST race-condition issue, but this may be expected, in which case I don't know what to do about it.

I will rebase and push the branch from time to time. Obviously let me know if you have issues.

**Comment by xianyuIC** [comment]
Sorry for the late reply, I used the latest neovintageous, but the situation still exists, just for a shorter time. I can still see from my eyes that when I first opened the file, it was in insert mode, and then it took about 0.5s to change to visual mode.

There is a common use: I just open the file to press "G" to reach the bottom of the file, the result becomes input G, but did not reach the bottom of the file, must wait for 0.5s,  wait for sublime automatically into visual mode, then press G , it can sucessful.

As described earlier, this only happens when you just open your first file with sublime. If a file already exists in sublime, this won't happen when you open a second file again.

**Comment by gerardroche** [comment]
The fix I applied only exists in the `default-visual-mode` branch https://github.com/NeoVintageous/NeoVintageous/commit/518a553306e4cb2886734aa0d4d33612fdf1292d 

The fix is not really a fix but a hack. It uses `on_activated_async()` instead of `on_activated()`, which mostly fixes the issue, but there is still a race-condition. I can't apply this to the main branch. I don't see any way to fix this at the moment. This is an ST issue/limitation.
